### PR TITLE
Add partition support for pulsar

### PR DESCRIPTION
**Pull request by bingo1229**
# Description

Add Partitioning support for message ordering guarantees of same key.

## Issue reference

Support additional metadata for Apache Pulsar Client Producer option to set Key on the Pulsar Message #2608

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Add partitionKey for publish metadata.
* [X] Add subscribeType, processMode for subscribe metadata.

### CODE REVIEW

**Comment by berndverst** [comment]
Can you write an integration test here (if necessary spinning up docker containers with multiple partitions) to validate that specifying the partition key really works end to end?

Your code has to be added here - https://github.com/dapr/components-contrib/tree/master/tests/certification/pubsub/pulsar

**Comment by oops-oom** [comment]
> Can you write an integration test here (if necessary spinning up docker containers with multiple partitions) to validate that specifying the partition key really works end to end?
> 
> Your code has to be added here - https://github.com/dapr/components-contrib/tree/master/tests/certification/pubsub/pulsar

@berndverst integration test has been added.

File: pubsub/pulsar/pulsar.go [hunk 1]
```diff
@@ -318,22 +334,73 @@ func parsePublishMetadata(req *pubsub.PublishRequest, schema schemaMetadata) (
...
-		msg.DeliverAt, err = time.Parse(time.RFC3339, val)
-		if err != nil {
-			return nil, err
-		}
-	}
-	if val, ok := req.Metadata[deliverAfter]; ok {
-		msg.DeliverAfter, err = time.ParseDuration(val)
-		if err != nil {
-			return nil, err
+	for name, value := range req.Metadata {
```
Comment (ItalyPaleAle) [inline pubsub/pulsar/pulsar.go:337]: Before the switch, please also add a check to make sure that `value != ""`.

Comment (ItalyPaleAle) [reply pubsub/pulsar/pulsar.go:337]: Something like:

```go
	for name, value := range req.Metadata {
		if value == "" {
			continue
		}

		switch name {
```

File: pubsub/pulsar/pulsar.go [hunk 2]
```diff
@@ -318,22 +334,73 @@ func parsePublishMetadata(req *pubsub.PublishRequest, schema schemaMetadata) (
...
 	return msg, nil
 }
 
+// default: shared
+func getSubscribeType(req pubsub.SubscribeRequest) pulsar.SubscriptionType {
+	metadata := req.Metadata
+
+	var subsType pulsar.SubscriptionType
+
+	subsTypeStr, _ := metadata[subscribeTypeKey]
```
Comment (ItalyPaleAle) [inline pubsub/pulsar/pulsar.go:368]: Should we lowercase this?

```suggestion
	subsTypeStr := strings.ToLower(metadata[subscribeTypeKey])
```

File: pubsub/pulsar/pulsar.go [hunk 3]
```diff
@@ -318,22 +334,73 @@ func parsePublishMetadata(req *pubsub.PublishRequest, schema schemaMetadata) (
...
+	return subsType
+}
+
+// true: async,default  false: sync
+func getProcessMode(req pubsub.SubscribeRequest) EnableAsyncProcess {
+	metadata := req.Metadata
+
+	var processMode EnableAsyncProcess
+
+	processModeStr, _ := metadata[processModeKey]
```
Comment (ItalyPaleAle) [inline pubsub/pulsar/pulsar.go:391]: Here too?


```suggestion
	processModeStr := strings.ToLower(metadata[processModeKey])
```

File: pubsub/pulsar/pulsar.go [hunk 4]
```diff
@@ -373,28 +440,36 @@ func (p *Pulsar) Subscribe(ctx context.Context, req pubsub.SubscribeRequest, han
...
+	originTopic := req.Topic
 	var err error
 	for {
 		select {
 		case msg := <-consumer.Chan():
-			// Go routine to handle multiple messages at once.
-			p.wg.Add(1)
-			go func(msg pulsar.ConsumerMessage) {
-				defer p.wg.Done()
+			if getProcessMode(req) {
```
Comment (ItalyPaleAle) [inline pubsub/pulsar/pulsar.go:457]: I think returning a boolean from this method is not very clear. Perhaps consider something like `if processMode(req) == async`

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Thanks for the PR!! 🙌

**Review by ItalyPaleAle — DISMISSED** [review]

File: pubsub/pulsar/pulsar.go [hunk 5]
```diff
@@ -373,28 +423,36 @@ func (p *Pulsar) Subscribe(ctx context.Context, req pubsub.SubscribeRequest, han
...
+	originTopic := req.Topic
 	var err error
 	for {
 		select {
 		case msg := <-consumer.Chan():
-			// Go routine to handle multiple messages at once.
-			p.wg.Add(1)
-			go func(msg pulsar.ConsumerMessage) {
-				defer p.wg.Done()
+			if strings.ToLower(req.Metadata[processModeKey]) == processModeSync {
```
Comment (ItalyPaleAle) [inline pubsub/pulsar/pulsar.go:440]:
```suggestion
			if strings.ToLower(req.Metadata[processModeKey]) == processModeSync { //nolint:gocritic
```

**Pull request merged**